bbid.py

```diff
@@ -38,32 +44,37 @@ def download(url,output_dir):
 			print('FAIL: Image is a duplicate of ' + image_md5s[md5_key] + ', not saving ' + filename)
 			return
 
-		image_md5s[md5_key] = filename
+		nb_files = len(os.listdir(output_dir))
+		if nb_files >= limit:
+			return
 
+		image_md5s[md5_key] = filename
 		imagefile=open(os.path.join(output_dir, filename),'wb')
 		imagefile.write(image)
 		imagefile.close()
-		print("OK: " + filename)
+		print('OK ("%s" %d/%d): %s' % (search_id, nb_files+1, limit, filename))
 		tried_urls.append(url)
-	except Exception as e:
-		print("FAIL: " + filename)
+	except urllib.error.HTTPError as e:
+		print("FAIL (HTTPError): %s: %s" % (filename, e))
+	except urllib.error.URLError as e:
+		print("FAIL (URLError): %s: %s" % (filename, e))
+	except socket.timeout as e:
+		print("FAIL (timeout): %s: %s" % (filename, e))
+
 	finally:
-		in_progress.remove(filename)
+		in_progress[search_id].remove(filename)
 		pool_sema.release()
 
 def fetch_images_from_keyword(keyword,output_dir):
 	current = 1
-	last = ''
-	while True:
+	links = []
+	while len(links)<=limit*2: # We except at last 50% of valid images.
```

Comment:
Changes inside this block seem wrong on so many levels. Do I need to explain? Please return it to previous state, which was more cautious, less thread processing heavy and simpler.